### PR TITLE
SharePoint: Metadata

### PR DESCRIPTION
**Pull request by dynaway-kmi**
Read and write metadata of a list item

### CODE REVIEW

File: Modules/System/SharePoint/src/helpers/SharePointUriBuilder.Codeunit.al [hunk 1]
```diff
@@ -104,25 +110,37 @@ codeunit 9110 "SharePoint Uri Builder"
     procedure GetUri(): Text
     var
         FullUri: Text;
+        ParameterName, ParameterValue : Text;
     begin
         FullUri := UriLbl + Uri + '/';
         FullUri := FullUri.Replace('{server_name}', ServerName.TrimStart('/').TrimEnd('/')).Replace('{namespace}', Namespace.TrimStart('/').TrimEnd('/'));
+        if QueryParameters.Count() > 0 then begin
```
Comment (pri-kise) [inline Modules/System/SharePoint/src/helpers/SharePointUriBuilder.Codeunit.al:117]: Why aren't you using the URI Builder Codeunit, that supports QueryParameters?
https://github.com/microsoft/ALAppExtensions/blob/6c169363defaf6193dac3cf5f00bba83b06307de/Modules/System/URI/src/UriBuilder.Codeunit.al#L189

Please make sure that there is an dependency to the URI Module. https://github.com/microsoft/ALAppExtensions/blob/main/Modules/System/URI/app.json

File: Modules/System/SharePoint/src/model/listitem/SharePointListItem.Codeunit.al [hunk 2]
```diff
@@ -88,4 +90,9 @@ codeunit 9103 "SharePoint List Item"
             SharePointListItem."List Id" := SharePointUriBuilder.GetMethodParameter('Lists').Substring(6, 36);
         end;
     end;
+
+    [IntegrationEvent(false, false)]
```
Comment (pri-kise) [inline Modules/System/SharePoint/src/model/listitem/SharePointListItem.Codeunit.al:94]: Same for this event. 
Events of Internal Objects can't subscribed from other apps.

File: Modules/System/SharePoint/src/SharePointClientImpl.Codeunit.al [hunk 3]
```diff
@@ -589,6 +596,68 @@ codeunit 9101 "SharePoint Client Impl."
         SharePointFileParser.ParseSingleReturnValue(Result, SharePointFile);
         exit(true);
     end;
+
+    procedure UpdateListItemMetaDataField(ListTitle: Text; ItemId: Integer; ListItemEntityTypeFullName: Text; FieldName: Text; FieldValue: Text): Boolean
```
Comment (pri-kise) [inline Modules/System/SharePoint/src/SharePointClientImpl.Codeunit.al:600]: If I understand this correctly then this procedures is using the API v1.
https://learn.microsoft.com/en-us/sharepoint/dev/sp-add-ins/working-with-folders-and-files-with-rest
Does anybody know if  all methods are using the API v1 or if there are methods, that are using the API v2 (graph API)?
Maybe we can document which procedures we are using by referencing online documentation of the API, if documention is available.

File: Modules/System/SharePoint/src/helpers/SharePointRequestHelper.Codeunit.al [hunk 4]
```diff
@@ -70,12 +84,11 @@ codeunit 9109 "SharePoint Request Helper"
...
                 Headers.Add('Content-Type', SharePointHttpContent.GetContentType());
 
-            Headers.Add('X-RequestDigest', SharePointHttpContent.GetRequestDigest());
             RequestMessage.Content(HttpContent);
         end;
     end;
 
-    [NonDebuggable]
+    //[NonDebuggable]
     local procedure SendRequest(HttpRequestMessage: HttpRequestMessage) OperationResponse: Codeunit "SharePoint Operation Response"
```
Comment (pri-kise) [inline Modules/System/SharePoint/src/helpers/SharePointRequestHelper.Codeunit.al:92]: I guess this change was only made for debugging purpose and should ber reverted?

File: Modules/System/SharePoint/src/SharePointEvents.Codeunit.al [hunk 5]
```diff
@@ -0,0 +1,12 @@
+codeunit 9112 "SharePoint Events"
+{
+    [IntegrationEvent(false, false)]
+    procedure ProcessSharePointFileMetadata(Metadata: JsonToken; SharePointFile: Record "SharePoint File" temporary)
```
Comment (pri-kise) [inline Modules/System/SharePoint/src/SharePointEvents.Codeunit.al:4]: This event is missing some documentation.
The event should be an internal procedure.
And the codeunit should get the Access= Public property added.

File: Modules/System/SharePoint/src/SharePointEvents.Codeunit.al [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+    begin
+    end;
+
+    [IntegrationEvent(false, false)]
+    procedure ProcessSharePointListItemMetadata(Metadata: JsonToken; SharePointListItem: Record "SharePoint List Item" temporary)
```
Comment (pri-kise) [inline Modules/System/SharePoint/src/SharePointEvents.Codeunit.al:9]: Could you please add some documentation.
And shouldn't the SharePointListItem be a var parameter?

File: Modules/System/SharePoint/src/SharePointClient.Codeunit.al [hunk 7]
```diff
@@ -315,7 +340,48 @@ codeunit 9100 "SharePoint Client"
...
+    /// Adds a file to specific folder.
+    /// </summary>
+    /// <remarks>Does not require UI interaction.</remarks>
+    /// <param name="ServerRelativeUrl">URL of the parent folder.</param>
+    /// <param name="FileName">File name to use on SharePoint.</param>
+    /// <param name="FileInStream">File stream to upload.</param>
+    /// <param name="SharePointFile">Collection of the result (temporary record). Always one element.</param>
+    /// <param name="ListAllFields">Include metadata in results.</param>
+    /// <returns>True if the operation was successful; otherwise - false.</returns>
+    procedure AddFileToFolder(ServerRelativeUrl: Text; FileName: Text; var FileInStream: InStream; var SharePointFile: Record "SharePoint File" temporary; ListAllFields: Boolean): Boolean
```
Comment (pri-kise) [inline Modules/System/SharePoint/src/SharePointClient.Codeunit.al:356]: If I understand your code correctly then this function can raise the event `ProcessSharePointFileMetadata`?

Could you please update the documentation to include a raises information for the events, that migth be raised by this procedure?

**Comment by JesperSchulz** [comment]
Let's try to get this ready for release with 2023 Wave 2.

**Review by JesperSchulz — APPROVED** [review]

**Comment by JesperSchulz** [comment]
Let's ship this with 2023 Wave 2 :-)

**Pull request merged**